src/err/err_state.rs

```diff
@@ -1,26 +1,49 @@
 use crate::{
-    exceptions::{PyBaseException, PyTypeError},
+    exceptions::PyBaseException,
     ffi,
     types::{PyTraceback, PyType},
-    AsPyPointer, IntoPy, IntoPyPointer, Py, PyObject, Python,
+    AsPyPointer, IntoPy, IntoPyPointer, Py, PyAny, PyObject, Python,
 };
 
 #[derive(Clone)]
 pub(crate) struct PyErrStateNormalized {
+    #[cfg(not(Py_3_12))]
     pub ptype: Py<PyType>,
     pub pvalue: Py<PyBaseException>,
+    #[cfg(not(Py_3_12))]
     pub ptraceback: Option<Py<PyTraceback>>,
 }
 
+#[cfg(not(Py_3_12))]
+impl PyErrStateNormalized {
+    pub(crate) fn ptype<'py>(&'py self, py: Python<'py>) -> &'py PyType {
+        self.ptype.as_ref(py)
+    }
+
+    pub(crate) fn ptraceback<'py>(&'py self, py: Python<'py>) -> Option<&'py PyTraceback> {
+        self.ptraceback
+            .as_ref()
+            .map(|traceback| traceback.as_ref(py))
+    }
+}
+
+#[cfg(Py_3_12)]
+impl PyErrStateNormalized {
+    pub(crate) fn ptype<'py>(&'py self, py: Python<'py>) -> &'py PyType {
+        self.pvalue.as_ref(py).get_type()
+    }
+
+    pub(crate) fn ptraceback<'py>(&'py self, py: Python<'py>) -> Option<&'py PyTraceback> {
+        unsafe { py.from_owned_ptr_or_opt(ffi::PyException_GetTraceback(self.pvalue.as_ptr())) }
+    }
+}
+
+pub(crate) type PyErrStateLazyFn =
+    dyn for<'py> FnOnce(Python<'py>) -> (PyObject, PyObject) + Send + Sync;
```

Comment:
What is this `(PyObject, PyObject)`? Can this have a comment, or maybe could it return a struct instead?

Reply:
Completely agreed, I've changed it to a `struct` in a second commit.